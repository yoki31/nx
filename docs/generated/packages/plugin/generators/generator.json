{
  "name": "generator",
  "factory": "./src/generators/generator/generator#generatorGeneratorInternal",
  "schema": {
    "$schema": "https://json-schema.org/schema",
    "cli": "nx",
    "$id": "NxPluginGenerator",
    "title": "Create a Generator for an Nx Plugin",
    "description": "Create a Generator for an Nx Plugin.",
    "type": "object",
    "examples": [
      {
        "command": "nx g generator my-generator --project=my-plugin",
        "description": "Generate `libs/my-plugin/src/generators/my-generator`"
      }
    ],
    "properties": {
      "name": {
        "type": "string",
        "description": "Generator name.",
        "$default": { "$source": "argv", "index": 0 },
        "x-prompt": "What name would you like to use for the generator?",
        "x-priority": "important"
      },
      "directory": {
        "type": "string",
        "description": "The directory at which to create the generator file. When `--nameAndDirectoryFormat=as-provided`, it will be relative to the current working directory. Otherwise, it will be relative to the workspace root.",
        "aliases": ["dir"],
        "x-priority": "important"
      },
      "description": {
        "type": "string",
        "description": "Generator description.",
        "x-priority": "important"
      },
      "unitTestRunner": {
        "type": "string",
        "enum": ["jest", "none"],
        "description": "Test runner to use for unit tests.",
        "default": "jest"
      },
      "skipLintChecks": {
        "type": "boolean",
        "default": false,
        "description": "Do not add an eslint configuration for plugin json files."
      },
      "project": {
        "type": "string",
        "description": "The name of the project.",
        "alias": "p",
        "$default": { "$source": "projectName" },
        "x-deprecated": "Provide the `directory` option instead and use the `as-provided` format. The project will be determined from the directory provided. It will be removed in Nx v20."
      },
      "skipFormat": {
        "type": "boolean",
        "default": false,
        "description": "Do not format files with prettier.",
        "x-priority": "internal"
      },
      "nameAndDirectoryFormat": {
        "description": "Whether to generate the generator in the directory as provided, relative to the current working directory and ignoring the project (`as-provided`) or generate it using the project and directory relative to the workspace root (`derived`).",
        "type": "string",
        "enum": ["as-provided", "derived"]
      }
    },
    "required": ["name"],
    "additionalProperties": false,
    "presets": []
  },
  "description": "Create a generator for an Nx Plugin.",
  "implementation": "/packages/plugin/src/generators/generator/generator#generatorGeneratorInternal.ts",
  "aliases": [],
  "hidden": false,
  "path": "/packages/plugin/src/generators/generator/schema.json",
  "type": "generator"
}
