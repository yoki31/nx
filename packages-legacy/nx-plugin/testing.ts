export * from '@nx/plugin/testing';
