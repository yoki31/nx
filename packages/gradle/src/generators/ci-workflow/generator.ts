import {
  Tree,
  names,
  generateFiles,
  getPackageManagerCommand,
  NxJsonConfiguration,
  formatFiles,
  detectPackageManager,
  readNxJson,
  readJson,
} from '@nx/devkit';
import { join } from 'path';
import { getNxCloudUrl, isNxCloudUsed } from 'nx/src/utils/nx-cloud-utils';
import { deduceDefaultBase } from 'nx/src/utils/default-base';

function getCiCommands(ci: Schema['ci'], mainBranch: string): string[] {
  switch (ci) {
    case 'circleci': {
      return [`./nx affected --base=$NX_BASE --head=$NX_HEAD -t test build`];
    }
    default: {
      return [`./nx affected -t test build`];
    }
  }
}

export interface Schema {
  name: string;
  ci: 'github' | 'circleci';
  packageManager?: null;
  commands?: string[];
}

export async function ciWorkflowGenerator(tree: Tree, schema: Schema) {
  const ci = schema.ci;

  const options = getTemplateData(tree, schema);
  generateFiles(tree, join(__dirname, 'files', ci), '', options);
  await formatFiles(tree);
}

interface Substitutes {
  mainBranch: string;
  workflowName: string;
  workflowFileName: string;
  packageManager: string;
  packageManagerPrefix: string;
  commands: string[];
  nxCloudHost: string;
}

function getTemplateData(tree: Tree, options: Schema): Substitutes {
  const { name: workflowName, fileName: workflowFileName } = names(
    options.name
  );
  const packageManager = detectPackageManager();
  const { exec: packageManagerPrefix } =
    getPackageManagerCommand(packageManager);

  let nxCloudHost: string = 'nx.app';
  try {
    const nxCloudUrl = getNxCloudUrl(readJson(tree, 'nx.json'));
    nxCloudHost = new URL(nxCloudUrl).host;
  } catch {}

  const mainBranch = deduceDefaultBase();

  const commands = options.commands ?? getCiCommands(options.ci, mainBranch);

  return {
    workflowName,
    workflowFileName,
    packageManager,
    packageManagerPrefix,
    commands,
    mainBranch,
    nxCloudHost,
  };
}

export default ciWorkflowGenerator;
