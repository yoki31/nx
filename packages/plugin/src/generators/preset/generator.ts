import {
  formatFiles,
  GeneratorCallback,
  names,
  runTasksInSerial,
  Tree,
  updateJson,
} from '@nx/devkit';
import { Linter } from '@nx/eslint';
import { PackageJson } from 'nx/src/utils/package-json';
import { pluginGenerator } from '../plugin/plugin';
import { PresetGeneratorSchema } from './schema';
import createPackageGenerator from '../create-package/create-package';

export default async function (tree: Tree, options: PresetGeneratorSchema) {
  const tasks: GeneratorCallback[] = [];
  const pluginProjectName = names(
    options.pluginName.includes('/')
      ? options.pluginName.split('/')[1]
      : options.pluginName
  ).fileName;
  options.createPackageName =
    options.createPackageName === 'false' // for command line in e2e, it is passed as a string
      ? undefined
      : options.createPackageName;
  const pluginTask = await pluginGenerator(tree, {
    compiler: 'tsc',
    linter: Linter.EsLint,
    name: pluginProjectName,
    skipFormat: true,
    unitTestRunner: 'jest',
    importPath: options.pluginName,
    e2eTestRunner: 'jest',
    publishable: true,
    // when creating a CLI package, the plugin will be in the packages folder
    directory:
      options.createPackageName && options.createPackageName !== 'false'
        ? 'packages'
        : undefined,
    rootProject: options.createPackageName ? false : true,
  });
  tasks.push(pluginTask);

  moveNxPluginToDevDeps(tree);

  if (options.createPackageName) {
    const e2eProject = `${options.pluginName}-e2e`;
    const cliTask = await createPackageGenerator(tree, {
      directory: 'packages',
      name: options.createPackageName,
      e2eProject: e2eProject,
      project: options.pluginName,
      skipFormat: true,
      unitTestRunner: 'jest',
      linter: Linter.EsLint,
      compiler: 'tsc',
    });
    tasks.push(cliTask);
  }

  await formatFiles(tree);

  return runTasksInSerial(...tasks);
}
function moveNxPluginToDevDeps(tree: Tree) {
  updateJson<PackageJson>(tree, 'package.json', (json) => {
    const nxPluginEntry = json.dependencies['@nx/plugin'];
    delete json.dependencies['@nx/plugin'];
    json.devDependencies['@nx/plugin'] = nxPluginEntry;
    return json;
  });
}
