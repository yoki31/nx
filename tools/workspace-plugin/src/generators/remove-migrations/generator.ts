import {
  formatFiles,
  readJson,
  Tree,
  updateJson,
  visitNotIgnoredFiles,
} from '@nx/devkit';
import { basename, dirname, join } from 'path';
import { lte, major } from 'semver';

export interface RemoveMigrationsGeneratorSchema {
  v: number;
}

export async function removeMigrationsGenerator(
  tree: Tree,
  { v }: RemoveMigrationsGeneratorSchema
) {
  visitNotIgnoredFiles(tree, '', (path) => {
    // Ignore angular migrations because angular needs to support migrations until LTS support is dropped
    if (['packages/angular/package.json'].includes(path)) {
      return;
    }

    if (basename(path) !== 'package.json') {
      return;
    }

    const packageJson = readJson(tree, path);

    const migrations =
      packageJson?.['nx-migrations']?.migrations ??
      packageJson?.['ng-update']?.migrations;
    if (!migrations) {
      return;
    }

    if (migrations.startsWith('.')) {
      const migrationsPath = join(dirname(path), migrations);

      updateJson(tree, migrationsPath, (migrationsJson) => {
        const { generators, packageJsonUpdates } = migrationsJson;
        for (const [migrationName, m] of Object.entries(generators)) {
          if (major(m.version) < v) {
            const implFile = getImplFile(tree, migrationsPath, m);

            const specFile = implFile.replace('.ts', '.spec.ts');
            const snapshotsPath = join(dirname(specFile), '__snapshots__');
            try {
              tree.delete(implFile);
              tree.delete(specFile);
              tree.delete(snapshotsPath);
            } catch (e) {
              console.log(e);
            }

            delete migrationsJson.generators[migrationName];
          }
        }

        for (const [updateName, packageJsonUpdate] of Object.entries(
          packageJsonUpdates ?? {}
        )) {
          if (major(packageJsonUpdate.version) < v) {
            delete packageJsonUpdates[updateName];
          }
        }

        return migrationsJson;
      });
    }
  });
  await formatFiles(tree);
}

export default removeMigrationsGenerator;

function getImplFile(
  tree: Tree,
  migrationsFilePath: string,
  {
    implementation,
    factory,
  }: {
    implementation?: string;
    factory?: string;
  }
) {
  const rawPath = implementation ?? factory;

  const fullPath = join(dirname(migrationsFilePath), rawPath);

  if (tree.exists(fullPath)) {
    return fullPath;
  }
  if (tree.exists(fullPath + '.ts')) {
    return fullPath + '.ts';
  }

  return fullPath;
}
