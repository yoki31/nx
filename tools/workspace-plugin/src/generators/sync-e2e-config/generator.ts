import {
  formatFiles,
  getProjects,
  globAsync,
  Tree,
  updateProjectConfiguration,
} from '@nx/devkit';
import { relative } from 'node:path';
import { join } from 'path';

export async function syncE2eConfigsGenerator(tree: Tree) {
  for (const [projName, projectConfig] of getProjects(tree)) {
    const { root, targets } = projectConfig;
    if (!root.startsWith('e2e/')) {
      continue;
    }

    const testFiles = await globAsync(tree, [join(root, `**/*.test.ts`)]);

    for (const testFile of testFiles) {
      const relativePath = relative(root, testFile);
      let target = (targets[`e2e-ci--${relativePath}`] ??= {});
      target.dependsOn = [
        'nx:build-native',
        '@nx/nx-source:populate-local-registry-storage',
      ];
      target.inputs = ['e2eInputs', '^production'];
    }
    updateProjectConfiguration(tree, projName, projectConfig);
  }

  await formatFiles(tree);
}

export default syncE2eConfigsGenerator;
