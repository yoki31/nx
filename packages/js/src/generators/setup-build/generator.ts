import {
  ensurePackage,
  formatFiles,
  type GeneratorCallback,
  joinPathFragments,
  readProjectConfiguration,
  runTasksInSerial,
  type Tree,
  updateProjectConfiguration,
} from '@nx/devkit';
import { addSwcConfig } from '../../utils/swc/add-swc-config';
import { addSwcDependencies } from '../../utils/swc/add-swc-dependencies';
import { nxVersion } from '../../utils/versions';
import { SetupBuildGeneratorSchema } from './schema';
import { addBuildTargetDefaults } from '@nx/devkit/src/generators/add-build-target-defaults';

export async function setupBuildGenerator(
  tree: Tree,
  options: SetupBuildGeneratorSchema
): Promise<GeneratorCallback> {
  const tasks: GeneratorCallback[] = [];
  const project = readProjectConfiguration(tree, options.project);
  const buildTarget = options.buildTarget ?? 'build';
  const prevBuildOptions = project.targets?.[buildTarget]?.options;

  project.targets ??= {};

  let mainFile: string;
  if (prevBuildOptions?.main) {
    mainFile = prevBuildOptions.main;
  } else if (options.main) {
    mainFile = options.main;
  } else {
    const root = project.sourceRoot ?? project.root;
    for (const f of [
      joinPathFragments(root, 'main.ts'),
      joinPathFragments(root, 'main.js'),
      joinPathFragments(root, 'index.ts'),
      joinPathFragments(root, 'index.js'),
    ]) {
      if (tree.exists(f)) {
        mainFile = f;
        break;
      }
    }
  }
  if (!mainFile || !tree.exists(mainFile)) {
    throw new Error(
      `Cannot locate a main file for ${options.project}. Please specify one using --main=<file-path>.`
    );
  }

  let tsConfigFile: string;
  if (prevBuildOptions?.tsConfig) {
    tsConfigFile = prevBuildOptions.tsConfig;
  } else if (options.tsConfig) {
    tsConfigFile = options.tsConfig;
  } else {
    for (const f of [
      'tsconfig.lib.json',
      'tsconfig.app.json',
      'tsconfig.json',
    ]) {
      const candidate = joinPathFragments(project.root, f);
      if (tree.exists(candidate)) {
        tsConfigFile = candidate;
        break;
      }
    }
  }
  if (!tsConfigFile || !tree.exists(tsConfigFile)) {
    throw new Error(
      `Cannot locate a tsConfig file for ${options.project}. Please specify one using --tsConfig=<file-path>.`
    );
  }

  switch (options.bundler) {
    case 'vite': {
      const { viteConfigurationGenerator } = ensurePackage(
        '@nx/vite',
        nxVersion
      );
      const task = await viteConfigurationGenerator(tree, {
        buildTarget: options.buildTarget,
        project: options.project,
        newProject: true,
        uiFramework: 'none',
        includeVitest: false,
        includeLib: true,
        skipFormat: true,
      });
      tasks.push(task);
      break;
    }
    case 'esbuild': {
      const { configurationGenerator } = ensurePackage(
        '@nx/esbuild',
        nxVersion
      );
      const task = await configurationGenerator(tree, {
        main: mainFile,
        buildTarget: options.buildTarget,
        project: options.project,
        skipFormat: true,
        skipValidation: true,
      });
      tasks.push(task);
      break;
    }
    case 'rollup': {
      const { configurationGenerator } = ensurePackage('@nx/rollup', nxVersion);
      const task = await configurationGenerator(tree, {
        buildTarget: options.buildTarget,
        main: mainFile,
        tsConfig: tsConfigFile,
        project: options.project,
        compiler: 'tsc',
        format: ['cjs', 'esm'],
        skipFormat: true,
        skipValidation: true,
      });
      tasks.push(task);
      break;
    }
    case 'tsc': {
      addBuildTargetDefaults(tree, '@nx/js:tsc');

      const outputPath = joinPathFragments('dist', project.root);
      project.targets[buildTarget] = {
        executor: `@nx/js:tsc`,
        outputs: ['{options.outputPath}'],
        options: {
          outputPath,
          main: mainFile,
          tsConfig: tsConfigFile,
          assets: [],
        },
      };
      updateProjectConfiguration(tree, options.project, project);
      break;
    }
    case 'swc': {
      addBuildTargetDefaults(tree, '@nx/js:swc');

      const outputPath = joinPathFragments('dist', project.root);
      project.targets[buildTarget] = {
        executor: `@nx/js:swc`,
        outputs: ['{options.outputPath}'],
        options: {
          outputPath,
          main: mainFile,
          tsConfig: tsConfigFile,
          assets: [],
        },
      };
      updateProjectConfiguration(tree, options.project, project);
      addSwcDependencies(tree);
      addSwcConfig(tree, project.root, 'es6');
    }
  }

  await formatFiles(tree);

  return runTasksInSerial(...tasks);
}

export default setupBuildGenerator;
