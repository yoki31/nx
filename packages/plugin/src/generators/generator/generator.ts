import {
  formatFiles,
  GeneratorsJson,
  joinPathFragments,
  Tree,
  writeJson,
  generateFiles,
  names,
  readJson,
  readProjectConfiguration,
  updateJson,
} from '@nx/devkit';
import { PackageJson } from 'nx/src/utils/package-json';
import { hasGenerator } from '../../utils/has-generator';
import pluginLintCheckGenerator from '../lint-checks/generator';
import type { Schema } from './schema';
import { nxVersion } from '../../utils/versions';
import { determineArtifactNameAndDirectoryOptions } from '@nx/devkit/src/generators/artifact-name-and-directory-utils';
import { join, relative } from 'path';

type NormalizedSchema = Schema & {
  fileName: string;
  className: string;
  propertyName: string;
  projectRoot: string;
  projectSourceRoot: string;
};

async function normalizeOptions(
  tree: Tree,
  options: Schema
): Promise<NormalizedSchema> {
  const { project, fileName, artifactName, filePath, directory } =
    await determineArtifactNameAndDirectoryOptions(tree, {
      artifactType: 'generator',
      callingGenerator: '@nx/plugin:generator',
      name: options.name,
      nameAndDirectoryFormat: options.nameAndDirectoryFormat,
      project: options.project,
      directory: options.directory,
      fileName: 'generator',
      derivedDirectory: 'generators',
    });

  const { className, propertyName } = names(artifactName);

  const { root: projectRoot, sourceRoot: projectSourceRoot } =
    readProjectConfiguration(tree, project);

  let description: string;
  if (options.description) {
    description = options.description;
  } else {
    description = `${options.name} generator`;
  }

  return {
    ...options,
    project,
    directory,
    fileName,
    className,
    propertyName,
    description,
    projectRoot,
    projectSourceRoot,
  };
}

function addFiles(host: Tree, options: NormalizedSchema) {
  const indexPath = join(options.directory, 'files/src/index.ts.template');

  if (!host.exists(indexPath)) {
    host.write(indexPath, 'const variable = "<%= name %>";');
  }

  generateFiles(host, join(__dirname, './files/generator'), options.directory, {
    ...options,
    generatorFnName: `${options.propertyName}Generator`,
    schemaInterfaceName: `${options.className}GeneratorSchema`,
  });

  if (options.unitTestRunner === 'none') {
    host.delete(join(options.directory, `generator.spec.ts`));
  }
}

export async function createGeneratorsJson(
  host: Tree,
  projectRoot: string,
  projectName: string,
  skipLintChecks?: boolean,
  skipFormat?: boolean
) {
  updateJson<PackageJson>(
    host,
    joinPathFragments(projectRoot, 'package.json'),
    (json) => {
      json.generators ??= './generators.json';
      return json;
    }
  );
  writeJson<GeneratorsJson>(
    host,
    joinPathFragments(projectRoot, 'generators.json'),
    {
      generators: {},
    }
  );
  if (!skipLintChecks) {
    await pluginLintCheckGenerator(host, {
      projectName,
      skipFormat,
    });
  }
}

async function updateGeneratorJson(host: Tree, options: NormalizedSchema) {
  const packageJson = readJson<PackageJson>(
    host,
    joinPathFragments(options.projectRoot, 'package.json')
  );
  const packageJsonGenerators =
    packageJson.generators ?? packageJson.schematics;
  let generatorsPath = packageJsonGenerators
    ? joinPathFragments(options.projectRoot, packageJsonGenerators)
    : null;

  if (!generatorsPath) {
    generatorsPath = joinPathFragments(options.projectRoot, 'generators.json');
  }
  if (!host.exists(generatorsPath)) {
    await createGeneratorsJson(
      host,
      options.projectRoot,
      options.project,
      options.skipLintChecks,
      options.skipFormat
    );
  }
  // add dependencies
  updateJson<PackageJson>(
    host,
    joinPathFragments(options.projectRoot, 'package.json'),
    (json) => {
      json.dependencies = {
        '@nx/devkit': nxVersion,
        ...json.dependencies,
      };
      return json;
    }
  );

  updateJson<GeneratorsJson>(host, generatorsPath, (json) => {
    let generators = json.generators ?? json.schematics;
    generators = generators || {};
    generators[options.name] = {
      factory: `./${joinPathFragments(
        relative(options.projectRoot, options.directory),
        'generator'
      )}`,
      schema: `./${joinPathFragments(
        relative(options.projectRoot, options.directory),
        'schema.json'
      )}`,
      description: options.description,
    };
    // @todo(v17): Remove this, prop is defunct.
    if (options.name === 'preset') {
      generators[options.name]['x-use-standalone-layout'] = true;
    }
    json.generators = generators;
    return json;
  });
}

export async function generatorGenerator(tree: Tree, rawOptions: Schema) {
  await generatorGeneratorInternal(tree, {
    nameAndDirectoryFormat: 'derived',
    ...rawOptions,
  });
}

export async function generatorGeneratorInternal(host: Tree, schema: Schema) {
  const options = await normalizeOptions(host, schema);
  if (hasGenerator(host, options.project, options.name)) {
    throw new Error(`Generator ${options.name} already exists.`);
  }

  addFiles(host, options);

  await updateGeneratorJson(host, options);

  if (!options.skipFormat) {
    await formatFiles(host);
  }
}

export default generatorGenerator;
