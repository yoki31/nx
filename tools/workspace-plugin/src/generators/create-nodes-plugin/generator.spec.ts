import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';
import { addProjectConfiguration, Tree, writeJson } from '@nx/devkit';

import { generatorGenerator } from './generator';
import { setCwd } from '@nx/devkit/internal-testing-utils';

describe('create-nodes-plugin/generator generator', () => {
  let tree: Tree;

  beforeEach(() => {
    tree = createTreeWithEmptyWorkspace();

    addProjectConfiguration(tree, 'eslint', {
      root: 'packages/eslint',
      targets: {
        build: {},
      },
    });

    jest.spyOn(process, 'cwd').mockReturnValue('/virtual/packages/eslint');

    setCwd('packages/eslint');
  });

  it('should run successfully', async () => {
    writeJson(tree, 'packages/eslint/package.json', {});
    await generatorGenerator(tree);
    expect(
      tree.read('packages/eslint/src/plugins/plugin.ts').toString()
    ).toMatchSnapshot();
    expect(
      tree.read('packages/eslint/src/plugins/plugin.spec.ts').toString()
    ).toMatchSnapshot();
    expect(
      tree
        .read(
          'packages/eslint/src/migrations/update-17-2-0/add-eslint-plugin.ts'
        )
        .toString()
    ).toMatchSnapshot();
  });

  it('should add the plugin path to package.json exports', async () => {
    writeJson(tree, 'packages/eslint/package.json', {
      name: '@nx/eslint',
      version: '0.0.1',
      private: false,
      description: 'Some description',
      repository: {
        type: 'git',
        url: 'https://github.com/nrwl/nx.git',
        directory: 'packages/eslint',
      },
      keywords: ['Monorepo', 'eslint', 'Web', 'CLI'],
      main: './index',
      typings: './index.d.ts',
      author: 'Victor Savkin',
      license: 'MIT',
      bugs: {
        url: 'https://github.com/nrwl/nx/issues',
      },
      homepage: 'https://nx.dev',
      generators: './generators.json',
      executors: './executors.json',
      'ng-update': {
        requirements: {},
        migrations: './migrations.json',
      },
      dependencies: {
        '@nx/devkit': 'file:../devkit',
      },
      peerDependencies: {},
      publishConfig: {
        access: 'public',
      },
      exports: {
        '.': './index.js',
        './package.json': './package.json',
        './migrations.json': './migrations.json',
        './generators.json': './generators.json',
        './executors.json': './executors.json',
        './executors': './executors.js',
        './src/executors/*/schema.json': './src/executors/*/schema.json',
        './src/executors/*.impl': './src/executors/*.impl.js',
        './src/executors/*/compat': './src/executors/*/compat.js',
      },
    });
    await generatorGenerator(tree);
    expect(
      tree.read('packages/eslint/package.json', 'utf-8')
    ).toMatchSnapshot();
  });
});
