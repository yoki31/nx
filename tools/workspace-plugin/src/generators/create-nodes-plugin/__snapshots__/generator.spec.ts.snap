// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create-nodes-plugin/generator generator should add the plugin path to package.json exports 1`] = `
"{
  "name": "@nx/eslint",
  "version": "0.0.1",
  "private": false,
  "description": "Some description",
  "repository": {
    "type": "git",
    "url": "https://github.com/nrwl/nx.git",
    "directory": "packages/eslint"
  },
  "keywords": [
    "Monorepo",
    "eslint",
    "Web",
    "CLI"
  ],
  "main": "./index",
  "typings": "./index.d.ts",
  "author": "Victor Savkin",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/nrwl/nx/issues"
  },
  "homepage": "https://nx.dev",
  "generators": "./generators.json",
  "executors": "./executors.json",
  "ng-update": {
    "requirements": {},
    "migrations": "./migrations.json"
  },
  "dependencies": {
    "@nx/devkit": "file:../devkit"
  },
  "peerDependencies": {},
  "publishConfig": {
    "access": "public"
  },
  "exports": {
    ".": "./index.js",
    "./package.json": "./package.json",
    "./migrations.json": "./migrations.json",
    "./generators.json": "./generators.json",
    "./executors.json": "./executors.json",
    "./executors": "./executors.js",
    "./src/executors/*/schema.json": "./src/executors/*/schema.json",
    "./src/executors/*.impl": "./src/executors/*.impl.js",
    "./src/executors/*/compat": "./src/executors/*/compat.js",
    "./plugin": "./plugin.js"
  }
}
"
`;

exports[`create-nodes-plugin/generator generator should run successfully 1`] = `
"import {
  CreateNodes,
  CreateNodesContext,
  TargetConfiguration,
} from '@nx/devkit';
import { basename, dirname, extname, join, resolve } from 'path';
import { registerTsProject } from '@nx/js/src/internal';

import { getRootTsConfigPath } from '@nx/js';

import { readTargetDefaultsForTarget } from 'nx/src/project-graph/utils/project-configuration-utils';
import { getNamedInputs } from '@nx/devkit/src/utils/get-named-inputs';
import { readdirSync } from 'fs';

export interface EslintPluginOptions {
  targetName?: string;
}

export const createNodes: CreateNodes<EslintPluginOptions> = [
  '**/TODO',
  (configFilePath, options, context) => {
    const projectRoot = dirname(configFilePath);

    // Do not create a project if package.json and project.json isn't there.
    const siblingFiles = readdirSync(join(context.workspaceRoot, projectRoot));
    if (
      !siblingFiles.includes('package.json') &&
      !siblingFiles.includes('project.json')
    ) {
      return {};
    }

    options = normalizeOptions(options);

    return {
      projects: {
        [projectRoot]: {
          root: projectRoot,
          projectType: 'library',
          targets: buildEslintTargets(
            configFilePath,
            projectRoot,
            options,
            context
          ),
        },
      },
    };
  },
];

function buildEslintTargets(
  configFilePath: string,
  projectRoot: string,
  options: EslintPluginOptions,
  context: CreateNodesContext
) {
  const eslintConfig = getEslintConfig(configFilePath, context);

  const targetDefaults = readTargetDefaultsForTarget(
    options.targetName,
    context.nxJsonConfiguration.targetDefaults,
    'executorName'
  );

  const namedInputs = getNamedInputs(projectRoot, context);

  const targets: Record<string, TargetConfiguration<ExecutorOptions>> = {};

  const baseTargetConfig: TargetConfiguration<ExecutorOptions> = {
    executor: 'executorName',
    options: {
      config: configFilePath,
      ...targetDefaults?.options,
    },
  };

  targets[options.targetName] = {
    ...baseTargetConfig,
    cache: targetDefaults?.cache ?? true,
    inputs:
      targetDefaults?.inputs ?? 'production' in namedInputs
        ? ['default', '^production']
        : ['default', '^default'],
    outputs: targetDefaults?.outputs ?? getOutputs(projectRoot),
    options: {
      ...baseTargetConfig.options,
    },
  };

  return targets;
}

function getEslintConfig(
  configFilePath: string,
  context: CreateNodesContext
): any {
  const resolvedPath = join(context.workspaceRoot, configFilePath);

  let module: any;
  if (extname(configFilePath) === '.ts') {
    const tsConfigPath = getRootTsConfigPath();

    if (tsConfigPath) {
      const unregisterTsProject = registerTsProject(tsConfigPath);
      try {
        module = require(resolvedPath);
      } finally {
        unregisterTsProject();
      }
    } else {
      module = require(resolvedPath);
    }
  } else {
    module = require(resolvedPath);
  }
  return module.default ?? module;
}

function getOutputs(projectRoot: string): string[] {
  function getOutput(path: string): string {
    if (path.startsWith('..')) {
      return join('{workspaceRoot}', join(projectRoot, path));
    } else {
      return join('{projectRoot}', path);
    }
  }

  const outputs = [];

  return outputs;
}

function normalizeOptions(options: EslintPluginOptions): EslintPluginOptions {
  options ??= {};
  options.targetName ??= 'TODO';
  return options;
}
"
`;

exports[`create-nodes-plugin/generator generator should run successfully 2`] = `
"import { CreateNodesContext } from '@nx/devkit';

import { createNodes } from './plugin';

describe('@nx/eslint/plugin', () => {
  let createNodesFunction = createNodes[1];
  let context: CreateNodesContext;

  beforeEach(async () => {
    context = {
      nxJsonConfiguration: {
        namedInputs: {
          default: ['{projectRoot}/**/*'],
          production: ['!{projectRoot}/**/*.spec.ts'],
        },
      },
      workspaceRoot: '',
    };
  });

  afterEach(() => {
    jest.resetModules();
  });

  it('should create nodes', () => {
    mockEslintConfig({});
    const nodes = createNodesFunction(
      'TODO',
      {
        targetName: 'target',
      },
      context
    );

    expect(nodes).toMatchInlineSnapshot();
  });
});

function mockEslintConfig(config: any) {
  jest.mock(
    'TODO',
    () => ({
      default: config,
    }),
    {
      virtual: true,
    }
  );
}
"
`;

exports[`create-nodes-plugin/generator generator should run successfully 3`] = `
"import { formatFiles, getProjects, Tree } from '@nx/devkit';
import { createNodes } from '../../plugins/plugin';

import { createProjectRootMappingsFromProjectConfigurations } from 'nx/src/project-graph/utils/find-project-for-path';
import { replaceProjectConfigurationsWithPlugin } from '@nx/devkit/src/utils/replace-project-configuration-with-plugin';

export default async function update(tree: Tree) {
  const proj = Object.fromEntries(getProjects(tree).entries());

  const rootMappings = createProjectRootMappingsFromProjectConfigurations(proj);

  await replaceProjectConfigurationsWithPlugin(
    tree,
    rootMappings,
    '@nx/eslint/plugin',
    createNodes,
    {
      targetName: 'TODO',
    }
  );

  await formatFiles(tree);
}
"
`;
